File: GenerateParantheses/GenerateParentheses.hs
generateParentheses :: Int -> [String]
generateParentheses = go "" 0
  where
    go :: String -> Int -> Int -> [String]
    go "" o n
      | n > 0 = go ")" (o + 1) (n - 1)
      | otherwise = []
    go s o n = case (o > 0, n > 0) of
      (True, True) -> go (')' : s) (o + 1) (n - 1) ++ go ('(' : s) (o - 1) n
      (True, False) -> go ('(' : s) (o - 1) n
      (False, True) -> go (')' : s) (o + 1) (n - 1)
      _ -> [s]

File: GenerateParantheses/generate_parentheses.py
def generate_parentheses(n: int) -> list[str]:
    def go(s: str, o: int, m: int) -> list[str]:
        if not s:
            if m > 0:
                return go(")", o + 1, m - 1)
            return []
        if o > 0:
            if m > 0:
                return go(")" + s, o + 1, m - 1) + (go("(" + s, o - 1, m))
            else:
                return go("(" + s, o - 1, m)
        else:
            if m > 0:
                return go(")" + s, o + 1, m - 1)
        return [s]

    return go("", 0, n)
